### PR TITLE
Feature/interaction design

### PR DESCRIPTION
**Pull request by deepweeb**
#201
# Introduction
API was completed, but it is not optimized because there are still many methods that can be combined and improved.
These are the fundamental differences between the methods from the new API (left) and the old API (right):
image

(Click image to enlarge)

- `AddContactToChain `is changed to `addContact`
- `AddContactToDatabase `is removed and  joined into `addContact`
- `addContact`will take in both User and Acquaintance objects;
- `addContact` adds a new contact to your chain. This contact's genesis block must be in the database before this happens. In case of pairing, `addAcquaintanceMultichain `must be called first to add the genesis block of this contact.
- `addAcquaintanceMultichain`  adds the Acquaintance's multichain (the database) which already contain this person's genesis block, block per block like what `AddContactToDatabase` does.  

- `revokeBlock `(confusing name, this method is used to adjust the trust after revoking, not to revoke a block but it has this name) is removed and joined into `revokeContact `method.
- `revokeContactFromChain `is changed to `revokeContact`. It stays the same but will also adjust the trust value of the contact like what the `revokeBlock `method does. This reduces the amount of methods in the API

- `makeAcquaintanceObject `method is added (Thanks to @ymte ). This method will return an Acquaintance object with your whole database (including your genesis block) parsed into the multichain. This Acquaintance object is then given to the person you have paired with so he/she can add your multichain with `addAcquaintanceMultichain` and add you with their `addContact`method in case they want to add you to their chain (and vice versa). This way the process of pairing and adding will be more simple as oppose to the old API where you have to send your genesis block to the pairing partner and then your whole database.

- The Chain class is removed because it is redundant.

# Purpose
The API should be adjusted so that it is efficient and robust (less methods but should have no loop holes).

# Details
This adjustment is needed for the Interaction Design Module.

# Future Applicability
There might be future changes to the API until the latest release.

### CODE REVIEW

File: app/src/main/java/nl/tudelft/bbw/API.java [hunk 1]
```diff
@@ -46,15 +49,34 @@ public static void initializeAPI(User owner, Context context) throws HashExcepti
...
+        List<List<Block>> multichain = acquaintance.getMultichain();
+        if (multichain.isEmpty()) {
+            return;
+        }
+        for (List<Block> chain : multichain) {
+            for (Block block : chain) {
+                if (block.getBlockType() == BlockType.GENESIS) {
+                    blockController.createGenesis(block.getBlockOwner());
+                } else if (block.isRevoked()) {
+                    blockController.createRevokeBlock(block.getBlockOwner(), block.getContact());
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/API.java]: What if you want to add a block which already exists in the database?

File: app/src/main/java/nl/tudelft/bbw/API.java [hunk 2]
```diff
@@ -46,15 +49,34 @@ public static void initializeAPI(User owner, Context context) throws HashExcepti
...
+    public static void addAcquaintanceToChain(Acquaintance acquaintance, byte[] signature, byte[] message)
             throws HashException, BlockAlreadyExistsException {
-        blockController.addBlockToChain(contact, signature, message);
+
+        //Adding his database into your database (so you can look up his contacts)
+        List<List<Block>> multichain = acquaintance.getMultichain();
+        if (multichain.isEmpty()) {
+            return;
+        }
+        for (List<Block> chain : multichain) {
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/API.java]: Try to keep the methods in the API as small as possible, move the actual computing into one of the controllers, that's their function.

File: app/src/main/java/nl/tudelft/bbw/API.java [hunk 3]
```diff
@@ -136,20 +138,26 @@ public static void failedTransaction(Block block) {
...
-     * Method to update the trustValue of a block after revoking
+     * Create an acquintance object that you can send over the network
      *
-     * @param block   the specific block
+     * @return a new acquintance object
      */
-    public static void revokedBlock(Block block) {
-        Block updatedBlock = TrustController.revokeBlock(block);
-        blockController.updateTrustOfBlock(updatedBlock);
+    public static Acquaintance makeAcquintanceObject() {
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/API.java]: Move this into one of the controllers

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 4]
```diff
@@ -261,4 +278,10 @@ private boolean verifySignature(byte[] signature, byte[] message) {
         }
     }
 
+    //For debugging purpose
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Is it though? I've seen a call to it in API line 155.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java [hunk 5]
```diff
@@ -114,4 +115,10 @@
      * @return boolean, representing whether the blocks are equal
      */
     boolean verifyTrustworthiness(Block block) throws HashException;
+
+    /**
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java]: Is the BlockControllerInterface still used?

File: app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java [hunk 6]
```diff
@@ -114,4 +115,10 @@
...
+     * Remove a chain with an owner from the database
+     * @param owner the owner of the chain to remove
+     */
+    void removeChainFromDatabase(User owner);
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java]: Why would you delete something from the database?

File: app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java [hunk 7]
```diff
@@ -15,14 +16,14 @@
     /**
      * Properties of a Acquaintance
      */
-    private ArrayList<Chain> multichain;
+    private List<List<Block>> multichain;
```
Comment (nzarin) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: List<List<Block>> How should I read this? It looks a bit weird lol

File: app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java (hunk 6 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java]: this was created by mistake

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 8]
```diff
@@ -66,12 +76,12 @@ public final void setUp() throws HashException, BlockAlreadyExistsException {
...
-        final byte[] signature = ED25519.generateSignature(message, newUser.getPrivateKey());
-        list = API.getBlocks(newUser);
-        API.addContactToChain(owner, signature, message);
-        assertNotEquals(API.getBlocks(newUser), list);
+
+        final byte[] message = newUser.getPublicKey().getEncoded();
+        final byte[] signature = ED25519.generateSignature(message, owner.getPrivateKey());
+        list = API.getBlocks(owner);
+        API.addAcquaintanceToChain(newUser, signature, message);
+        assertNotEquals(API.getBlocks(owner), list);
```
Comment (nzarin) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: I still find the getBlocks not a clear name. For other teams that will use our API. They are interested in the users, right? Can't you rename it to 'getContacts' or something?

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 4 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: @ymte  Can you fix this comment? Thanks

File: app/src/main/java/nl/tudelft/bbw/API.java (hunk 3 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/API.java]: @ymte Can you look at this? Thanks

File: app/src/main/java/nl/tudelft/bbw/API.java (hunk 1 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/API.java]: Then it won't just be added.  BlockController should be handling these cases.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    /**
+     * Convert all blocks to a list of list of blocks.
+     * This works by looping all blocks, adding them to the current block list, and when
+     * encountering a new user add the block list to the multichain and then reset it.
+     * @param cursor the cursor resulting from the query
+     */
+    @Override
+    public void parse(Cursor cursor) {
```
Comment (nzarin) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: where is the cursor closed?

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 10]
```diff
@@ -47,13 +49,21 @@
     @Before
     public final void setUp() throws HashException, BlockAlreadyExistsException {
         EdDSAPrivateKey privateKey = ED25519.generatePrivateKey();
-        owner = new User("Jeff", "iban", ED25519.getPublicKey(privateKey));
+        owner = new Acquaintance("Jeff", "iban", ED25519.getPublicKey(privateKey), new ArrayList<List<Block>>());
```
Comment (nzarin) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: why do you use an arraylist here? I one of my previous comments you changed the arraylist into list for the multichain.

File: app/src/main/java/nl/tudelft/bbw/API.java (hunk 2 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/API.java]: This is fixed.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java (hunk 7 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: Initially, there is a class called Chain (extends ArrayList<Block) which if implemented will result in a List<Chain> instead of List<List<Block>> like this. This class was made to serve the purpose of making the code more observable. But the API & BlockControllers were implemented without the Chain class which result in these "ugly" objects. I cannot use Chain anymore because the above reason.

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 8 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Sounds good, renaming it now.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java (hunk 9 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: @ymte Can you take a look at this? thanks

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 10 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Actually when initialize the Acquaintance, there should be a database with at least a genesis block of this contact available. But since this is a test Acquaintance, we do not have any predefined database so to make an Acquaintance so I filled an empty multichain in. But if you look further to the next codes lines, the multichain is set to the one with a genesis block like it's supposed to be.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java (hunk 7 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: Also, thats why Chain class is removed because no ones use it.

Comment (nzarin) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: The chain class is still present in the UML. Can you please update the UML then?

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 11]
```diff
@@ -48,6 +48,15 @@ public BlockController(User chainOwner, Context context) {
     }
 
     /**
+     *Method to return the own chain
+     * @return the list of block forming the own chain of the user
+     */
+    public final List<Block> returnOwnChain()
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Why is this needed? You are everywhere aware which user you are right?

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 12]
```diff
@@ -261,4 +278,35 @@ private boolean verifySignature(byte[] signature, byte[] message) {
         }
     }
 
+    //For debugging purpose
+    public Database getDatabase()
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: When moving the computions into the controllers this getter becomes redundant.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 4 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: I have fixed it

Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Yes, the comment has been fixed

File: app/src/main/java/nl/tudelft/bbw/API.java (hunk 3 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/API.java]: I have looked at it and moved it; I have spotted the error

File: app/src/main/java/nl/tudelft/bbw/API.java (hunk 1 above)
Comment (ashay123) [inline app/src/main/java/nl/tudelft/bbw/API.java]: No, it will throw an error and there won't be any blocks added after.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java (hunk 5 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockControllerInterface.java]: The BlockControllerInterface has been removed as it is indeed not used anymore.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java (hunk 9 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: This happens in the ReadQuery.execute method

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 12 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: You are absolutely right. The method has been moved and the getDatabase method has been deleted.

File: app/src/main/java/nl/tudelft/bbw/API.java [hunk 13]
```diff
@@ -32,75 +40,105 @@ private API() {
...
      * @throws BlockAlreadyExistsException when adding a block results in an error
      */
-    public static void addContactToDatabase(User owner, User contact)
-            throws HashException, BlockAlreadyExistsException {
-        blockController.createKeyBlock(owner, contact);
+    public static void addAcquaintance(Acquaintance acquaintance)
+            throws HashException, BlockAlreadyExistsException, NoSuchAlgorithmException, InvalidKeyException, SignatureException {
+
+        final byte[] message = acquaintance.getPublicKey().getEncoded();
+        final byte[] signature = ED25519.generateSignature(message, blockController.getOwnUser().getPrivateKey());
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/API.java]: You have to generate the signature in the activity. In the API, the signature only gets verified.

File: app/src/main/java/nl/tudelft/bbw/API.java [hunk 14]
```diff
@@ -32,75 +40,105 @@ private API() {
...
-     * @param contact the contact
-     * @throws HashException               When creating a block results in an error
-     * @throws BlockAlreadyExistsException when adding a block results in an error
+     * @return name of the API user
      */
-    public static void revokeContactFromChain(User contact)
-            throws HashException, BlockAlreadyExistsException {
-        blockController.revokeBlockFromChain(contact);
+    public static String getMyName() {
+        return blockController.getOwnUser().getName();
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/API.java]: Why do you have so much getters; whilst you can just create one getter for a user, and get attributes from that user.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java (hunk 7 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: A list of lists is quite inefficient, a better solution would be a map.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Block.java [hunk 15]
```diff
@@ -280,7 +280,7 @@ public boolean verifyBlock(Object o) {
     @Override
     public String toString() {
         return "Block#" + getSequenceNumber() + "{owner=" + owner.getName() + ",contact="
-                + contact.getName() + "}";
+                + contact.getName() + (isRevoked()?",REVOKE":"") + "}";
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Block.java]: Check the checkstyle on this one

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 16]
```diff
@@ -48,6 +50,15 @@ public BlockController(User chainOwner, Context context) {
     }
 
     /**
+     *Method to return the own chain
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Run checkstyle

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 17]
```diff
@@ -123,6 +133,14 @@ public final void updateTrustOfBlock(Block block) {
     }
 
     /**
+     * Retrieve the owner of this app
+     * @return the owner of this app
+     */
+    public User getOwnUser() {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: In the activity, you always already have a user, so why would you need this function

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 18]
```diff
@@ -261,4 +279,39 @@ private boolean verifySignature(byte[] signature, byte[] message) {
...
+     * @throws BlockAlreadyExistsException
+     * @throws HashException
+     */
+    public void addMultichain(List<List<Block>> multichain) throws BlockAlreadyExistsException, HashException {
+        if (multichain.isEmpty()) {
+            return;
+        }
+        for (List<Block> chain : multichain) {
+            for (Block block : chain) {
+                if (block.getBlockType() == BlockType.GENESIS) {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: What happens if there is already a block in the database, and you want to add that same block to your chain. Currently, it will throw a runtime exception, while this should not be the case.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 19]
```diff
@@ -261,4 +279,39 @@ private boolean verifySignature(byte[] signature, byte[] message) {
...
 
+    /**
+     * Method to add the multichain (the pairing person database) into your database
+     *
+     * @param multichain given the multichain of the pairing partner
+     * @throws BlockAlreadyExistsException
+     * @throws HashException
+     */
+    public void addMultichain(List<List<Block>> multichain) throws BlockAlreadyExistsException, HashException {
+        if (multichain.isEmpty()) {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: This if statement could be negated, and the rest of the code could be thrown in it.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 20]
```diff
@@ -261,4 +279,39 @@ private boolean verifySignature(byte[] signature, byte[] message) {
...
     }
 
+    /**
+     * Method to add the multichain (the pairing person database) into your database
+     *
+     * @param multichain given the multichain of the pairing partner
+     * @throws BlockAlreadyExistsException
+     * @throws HashException
+     */
+    public void addMultichain(List<List<Block>> multichain) throws BlockAlreadyExistsException, HashException {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: This function is too large.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java [hunk 21]
```diff
@@ -261,4 +279,39 @@ private boolean verifySignature(byte[] signature, byte[] message) {
...
+                } else {
+                    this.createKeyBlock(block.getBlockOwner(), block.getContact());
+                }
+            }
+        }
+    }
+
+    /**
+     * Create an acquintance object that you can send over the network
+r    * @return a new acquaintance object
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: why is there a 'r' in here?

File: app/src/main/java/nl/tudelft/bbw/controller/TrustController.java [hunk 22]
```diff
@@ -63,13 +63,13 @@ public static Block verifiedIBAN(Block block) {
...
     /**
-     * revokeBlock
+     * revokeBlockTrust
      * Sets the block value to the revoked trust value
      *
      * @param block given block to revoke
      * @return block with the new trust value
      */
-    public static Block revokeBlock(Block block) {
+    public static Block revokeBlockTrust(Block block) {
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/TrustController.java]: Why is this method renamed; it does not follow the naming style of the other functions.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java (hunk 9 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: This function really is too large.

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 23]
```diff
@@ -46,14 +47,20 @@
      */
     @Before
     public final void setUp() throws HashException, BlockAlreadyExistsException {
+
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Why are these whitelines necessary?

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 24]
```diff
@@ -46,14 +47,20 @@
...
         EdDSAPrivateKey privateKey = ED25519.generatePrivateKey();
-        owner = new User("Jeff", "iban", ED25519.getPublicKey(privateKey));
-        owner.setPrivateKey(privateKey);
-        API.initializeAPI(owner, RuntimeEnvironment.application);
-        list = API.getBlocks(owner);
+        newUser = new Acquaintance("Nick", "iban2", ED25519.getPublicKey(privateKey), new ArrayList<List<Block>>());
 
-        privateKey = ED25519.generatePrivateKey();
-        newUser = new User("Nick", "iban2", ED25519.getPublicKey(privateKey));
+        ArrayList<List<Block>> newUserMultichain = new ArrayList<List<Block>>();
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Change it to: List<List<Block>> newUserMultichain = new ArrayList<List<Block>>();.
Since you also have the same implementation in the implementation

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 25]
```diff
@@ -46,14 +47,20 @@
...
+        ArrayList<Block> test = new ArrayList<Block>();
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Same as above

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 26]
```diff
@@ -66,12 +73,10 @@ public final void setUp() throws HashException, BlockAlreadyExistsException {
     @Test
     public final void addContactToChainTest() throws HashException, BlockAlreadyExistsException,
             NoSuchAlgorithmException, InvalidKeyException, SignatureException {
-        API.initializeAPI(newUser, RuntimeEnvironment.application);
-        final byte[] message = owner.getPublicKey().getEncoded();
-        final byte[] signature = ED25519.generateSignature(message, newUser.getPrivateKey());
-        list = API.getBlocks(newUser);
-        API.addContactToChain(owner, signature, message);
-        assertNotEquals(API.getBlocks(newUser), list);
+
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Do these whitelines improve readability?

File: app/src/test/java/nl/tudelft/bbw/APITest.java [hunk 27]
```diff
@@ -146,16 +128,22 @@ public final void failedTransactionTest() {
...
-    /**
-     * Test if trustValue of a block changes after revoking
-     */
+
     @Test
-    public final void revokedBlockTest() {
-        API.revokedBlock(list.get(0));
-        assertNotEquals(list.get(0).getTrustValue(), TrustValues.INITIALIZED);
+    public final void makeAcquaintanceTest() throws NoSuchAlgorithmException, InvalidKeyException, SignatureException, BlockAlreadyExistsException, HashException {
+
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Do these whitelines improve readability?

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 11 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: it is fixed. yes this method is not even used.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 17 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Yeah thats right it is removed now.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 18 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: It will throw an DatabaseException. I have made the method to catch this exception in the API and then do nothing about it as it is already added. I do not want to tamper with the block controller.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 19 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Fixed

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 21 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: probably a typo. It is fixed.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java (hunk 9 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: @ymte Can you take a look at dis.

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 23 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: They are used to divide the parts of a method for better analysis. It's more of individual coding style.
But some unnecessary lines are removed.

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 24 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: It is fixed.

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 25 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: It is fixed.

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 27 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Fixed

File: app/src/test/java/nl/tudelft/bbw/APITest.java (hunk 26 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/APITest.java]: Fixed.

File: app/src/main/java/nl/tudelft/bbw/controller/TrustController.java (hunk 22 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/TrustController.java]: All the API names will be renamed after the analysis process of the Interaction Design interviews is finished. Hang on.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 20 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: @ymte

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 16 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: I have ran it with `gradlew build connectedCheck check lint test` yet I see no problem here. You can try.

File: app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java (hunk 9 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/database/read/DatabaseToMultichainQuery.java]: Dis has been fixed.

File: app/src/main/java/nl/tudelft/bbw/controller/BlockController.java (hunk 20 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/controller/BlockController.java]: Optimized.

File: app/src/main/java/nl/tudelft/bbw/controller/TrustController.java (hunk 22 above)
Comment (humanjas) [inline app/src/main/java/nl/tudelft/bbw/controller/TrustController.java]: This is the TrustController, not the api class

**Comment by deepweeb** [comment]
@ymte Before you commit, DO run the tests.  Now I'm sitting here with this.
image
Fix it, thanks.

File: app/src/main/java/nl/tudelft/bbw/controller/TrustController.java (hunk 22 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/controller/TrustController.java]: This is restored back  to revokeBlock as it is not my class.

File: app/src/test/java/nl/tudelft/bbw/BlockChainAPITest.java [hunk 28]
```diff
@@ -82,52 +82,27 @@ public final void addContactToChainTest() throws HashException, BlockAlreadyExis
...
-    }
-
-    /**
-     * Test if adding to database yields a different chain
-     *
-     * @throws HashException               When creating a block results in an error
-     * @throws BlockAlreadyExistsException when adding a block results in an error
-     */
-    @Test
-    public final void addContactToDatabase() throws HashException, BlockAlreadyExistsException {
```
Comment (ashay123) [inline app/src/test/java/nl/tudelft/bbw/BlockChainAPITest.java]: Make sure that you also test whether the database indeed contains all the blocks which you added

**Comment by deepweeb** [comment]
@ashay123 when you have paired with B, you will be given an acquaintance object from B which contains the whole database of B. The GUI is used to present the contacts of B to you which contains C. Since the multichain contain the whole database of B it also contain the genesis block of C. So when you add the multichain of B you will also get the genesis block of C in your database which you can backtrace to after added C into the database. To add C you just need to add C genesis block to your chain.

**Comment by ymte** [comment]
The DatabaseToMultichain query has been fixed; it was because of splitting the function in two. Indeed the unittests were not properly run.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Block.java [hunk 29]
```diff
@@ -280,19 +280,36 @@ public boolean verifyBlock(Object o) {
...
     public String toString() {
         return "Block#" + getSequenceNumber() + "{owner=" + owner.getName() + ",contact="
-                + contact.getName() + "}";
+                + contact.getName() + (isRevoked() ? ",REVOKE" : "") + "}";
     }
 
     /**
+     * ToString method designed for Interaction Design Module
+     */
+    public String toString2() {
```
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Block.java]: this method is used for the interaction design, since it is not needed this method is removed too.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Block.java [hunk 30]
```diff
@@ -280,19 +280,36 @@ public boolean verifyBlock(Object o) {
...
+        if (this.getBlockType().equals(BlockType.GENESIS)) {
+            return "Block#" + getSequenceNumber() + " {contact="
+                    + getContactName() + ", iban="
+                    + getContactIban() + (isRevoked() ? ",REVOKE" : "") + ", Trust="
+                    + getTrustValue() + "} (My Own Block)";
```
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Block.java]: (My Own Block) is changed to (Genesis Block)

File: app/src/test/java/nl/tudelft/bbw/BlockChainAPITest.java (hunk 28 above)
Comment (deepweeb) [inline app/src/test/java/nl/tudelft/bbw/BlockChainAPITest.java]: I actually just copied your implementation of the tests of the previous API. But it is now changed with a better implementation.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java (hunk 7 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Acquaintance.java]: As @ymte stated and myself: I think that the map is not the ideal candidate in this situation because our multichain is used to just carry data. You don't really have to iterate through it or select from it.

**Comment by deepweeb** [comment]
ashay123 I actually got what you meant and have fixed this:
- method AddAcquaintanceMultichain is added which should be used to add the whole multichain right after pariting. 
-addAcquaintance is changed to addContact which then will accept both User and Acquaintance. This should only be used if the genesis blocks of the concerned contacts are already in the database (after the calling of  AddAcquaintanceMultichain for example).

**Comment by ashay123** [comment]
Alright, nice that you implemented my suggestions, approved.

File: app/src/main/java/nl/tudelft/bbw/blockchain/Block.java (hunk 15 above)
Comment (deepweeb) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Block.java]: I have ran it with gradlew build connectedCheck check lint test yet I see no problem here. You can try.

Comment (ymte) [inline app/src/main/java/nl/tudelft/bbw/blockchain/Block.java]: You are missing pmd and checkstyle!
